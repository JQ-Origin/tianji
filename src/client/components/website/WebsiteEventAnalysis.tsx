import React, { useMemo } from 'react';
import dayjs from 'dayjs';
import { Empty } from 'antd';
import { useTranslation } from '@i18next-toolkit/react';

import { trpc } from '@/api/trpc';
import { InsightQueryChart } from '@/components/insights/InsightQueryChart';
import { LoadingView } from '@/components/LoadingView';
import { useCurrentWorkspaceId } from '@/store/user';

const MAX_EVENT_SERIES = 5;

export const WebsiteEventAnalysis: React.FC<{
  websiteId: string;
}> = React.memo((props) => {
  const { websiteId } = props;
  const { t } = useTranslation();
  const workspaceId = useCurrentWorkspaceId();

  const { startAt, endAt } = useMemo(() => {
    const end = dayjs().endOf('hour');
    const start = end.subtract(24, 'hour');

    return {
      startAt: start.valueOf(),
      endAt: end.valueOf(),
    };
  }, [websiteId]);

  const { data: eventMetrics = [], isLoading: isLoadingEventMetrics } =
    trpc.website.metrics.useQuery(
      {
        workspaceId,
        websiteId,
        type: 'event',
        startAt,
        endAt,
      },
      {
        trpc: {
          context: {
            skipBatch: true,
          },
        },
      }
    );

  const metricsConfig = useMemo(() => {
    return eventMetrics
      .filter((item) => Boolean(item.x))
      .slice(0, MAX_EVENT_SERIES)
      .map((item) => ({
        name: String(item.x),
        math: 'events' as const,
        alias: String(item.x),
      }));
  }, [eventMetrics]);

  return (
    <div className="flex h-full flex-col gap-4">
      <LoadingView isLoading={isLoadingEventMetrics} className="flex-1">
        {metricsConfig.length === 0 ? (
          <div className="flex h-full items-center justify-center">
            <Empty description={t('No event data yet')} />
          </div>
        ) : (
          <InsightQueryChart
            className="h-[320px]"
            workspaceId={workspaceId}
            insightId={websiteId}
            insightType="website"
            metrics={metricsConfig}
            filters={[]}
            groups={[]}
            time={{
              startAt,
              endAt,
              unit: 'hour',
            }}
            chartType="bar"
          />
        )}
      </LoadingView>
    </div>
  );
});
WebsiteEventAnalysis.displayName = 'WebsiteEventAnalysis';
