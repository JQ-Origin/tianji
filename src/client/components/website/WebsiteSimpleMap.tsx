import React, { useMemo } from 'react';
import { trpc } from '../../api/trpc';
import { useCurrentWorkspaceId } from '../../store/user';
import { SimpleWorldMap } from '../SimpleWorldMap';

interface WebsiteSimpleMapProps {
  websiteId: string;
  startAt: number;
  endAt: number;
}

export const WebsiteSimpleMap: React.FC<WebsiteSimpleMapProps> = React.memo(
  (props) => {
    const { websiteId, startAt, endAt } = props;
    const workspaceId = useCurrentWorkspaceId();

    const { data: metrics = [] } = trpc.website.metrics.useQuery({
      workspaceId,
      websiteId,
      type: 'country',
      startAt,
      endAt,
    });

    const data = useMemo(() => {
      return metrics.map((m) => ({
        country: m.x ?? '',
        visitors: m.y,
      }));
    }, [metrics]);

    return <SimpleWorldMap data={data} />;
  }
);
WebsiteSimpleMap.displayName = 'WebsiteSimpleMap';
