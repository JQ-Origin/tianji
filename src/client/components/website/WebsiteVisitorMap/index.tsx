import React from 'react';
import { trpc } from '../../../api/trpc';
import { useCurrentWorkspaceId } from '../../../store/user';
import { useGlobalRangeDate } from '../../../hooks/useGlobalRangeDate';
import loadable from '@loadable/component';
import { Loading } from '../../Loading';
import { useGlobalConfig } from '../../../hooks/useConfig';

const VisitorLeafletMap = loadable(() =>
  import('./VisitorLeafletMap').then((m) => m.VisitorLeafletMap)
);

const VisitorLarkMap = loadable(() =>
  import('./VisitorLarkMap').then((m) => m.VisitorLarkMap)
);

function useMapType() {
  const { amapToken, mapboxToken } = useGlobalConfig();

  if (mapboxToken) {
    return 'Mapbox';
  } else if (amapToken) {
    return 'Gaode';
  } else {
    return 'Leaflet';
  }
}

interface WebsiteVisitorMapProps {
  websiteId: string;
  fullScreen?: boolean;
}
export const WebsiteVisitorMap: React.FC<WebsiteVisitorMapProps> = React.memo(
  (props) => {
    const workspaceId = useCurrentWorkspaceId();
    const { startDate, endDate } = useGlobalRangeDate();
    const startAt = startDate.valueOf();
    const endAt = endDate.valueOf();
    const { data } = trpc.website.geoStats.useQuery({
      workspaceId,
      websiteId: props.websiteId,
      startAt,
      endAt,
    });

    const mapType = useMapType();

    if (!data) {
      return <Loading />;
    }

    if (mapType === 'Mapbox' || mapType === 'Gaode') {
      return (
        <VisitorLarkMap
          mapType={mapType}
          data={data}
          fullScreen={props.fullScreen}
        />
      );
    }

    return <VisitorLeafletMap data={data} fullScreen={props.fullScreen} />;
  }
);
WebsiteVisitorMap.displayName = 'WebsiteVisitorMap';
