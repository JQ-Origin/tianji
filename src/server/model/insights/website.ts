import { z } from 'zod';
import { insightsQuerySchema } from '../../utils/schema.js';
import { prisma } from '../_client.js';
import { printSQL } from '../../utils/prisma.js';
import { Prisma } from '@prisma/client';
import { FilterInfoType, FilterInfoValue, getDateArray } from '@tianji/shared';
import { get } from 'lodash-es';
import { EVENT_TYPE } from '../../utils/const.js';
import { env } from '../../utils/env.js';
import { InsightsSqlBuilder } from './shared.js';

export class WebsiteInsightsSqlBuilder extends InsightsSqlBuilder {
  getTableName() {
    return 'WebsiteEvent';
  }

  buildSelectQueryArr() {
    const { metrics } = this.query;

    return metrics.map((item) => {
      if (item.math === 'events') {
        if (item.name === '$all_event') {
          return Prisma.sql`count(1) as "$all_event"`;
        }

        if (item.name === '$page_view') {
          return Prisma.sql`sum(case WHEN "WebsiteEvent"."eventName" is null AND "WebsiteEvent"."eventType" = ${EVENT_TYPE.pageView} THEN 1 ELSE 0 END) as "$page_view"`;
        }

        return Prisma.sql`sum(case WHEN "WebsiteEvent"."eventName" = ${item.name} THEN 1 ELSE 0 END) as ${Prisma.raw(`"${item.name}"`)}`;
      } else if (item.math === 'sessions') {
        if (item.name === '$all_event') {
          return Prisma.sql`count(distinct "sessionId") as "$all_event"`;
        }

        if (item.name === '$page_view') {
          return Prisma.sql`count(distinct case WHEN "WebsiteEvent"."eventName" is null AND "WebsiteEvent"."eventType" = ${EVENT_TYPE.pageView} THEN "sessionId" ELSE null END) as "$page_view"`;
        }

        return Prisma.sql`count(distinct case WHEN "WebsiteEvent"."eventName" = ${item.name} THEN "sessionId" END) as ${Prisma.raw(`"${item.name}"`)}`;
      }

      return null;
    });
  }

  buildGroupSelectQueryArr() {
    const { groups } = this.query;
    let groupSelectQueryArr: Prisma.Sql[] = [];
    if (groups.length > 0) {
      for (const g of groups) {
        if (!g.customGroups) {
          groupSelectQueryArr.push(
            Prisma.sql`${this.getValueField(g.type)} as "%${Prisma.raw(g.value)}"`
          );
        } else if (g.customGroups && g.customGroups.length > 0) {
          for (const cg of g.customGroups) {
            groupSelectQueryArr.push(
              Prisma.sql`${this.buildFilterQueryOperator(
                g.type,
                cg.filterOperator,
                cg.filterValue
              )} as "%${Prisma.raw(`${g.value}|${cg.filterOperator}|${cg.filterValue}`)}"`
            );
          }
        }
      }
    }
    return groupSelectQueryArr;
  }

  buildInnerJoinQuery() {
    const { filters, groups } = this.query;
    let innerJoinQuery = Prisma.empty;
    if (filters.length > 0 || groups.length > 0) {
      innerJoinQuery = Prisma.sql`INNER JOIN "WebsiteEventData" ON "WebsiteEvent"."id" = "WebsiteEventData"."websiteEventId"`;

      if (filters.length > 0) {
        innerJoinQuery = Prisma.sql`${innerJoinQuery} AND ${Prisma.join(
          filters.map((filter) =>
            this.buildFilterQueryOperator(
              filter.type,
              filter.operator,
              filter.value
            )
          ),
          ' AND '
        )}`;
      }

      if (groups.length > 0) {
        innerJoinQuery = Prisma.sql`${innerJoinQuery} AND ${Prisma.join(
          groups.map(
            (g) => Prisma.sql`"WebsiteEventData"."eventKey" = ${g.value}`
          ),
          ' OR '
        )}`;
      }
    }
    return innerJoinQuery;
  }

  buildWhereQueryArr() {
    const { insightId, time, metrics } = this.query;
    const { startAt, endAt } = time;

    return [
      // website id
      Prisma.sql`"WebsiteEvent"."websiteId" = ${insightId}`,

      // date
      this.buildDateRangeQuery('"WebsiteEvent"."createdAt"', startAt, endAt),

      // event name
      Prisma.join(
        metrics.map((item) => {
          if (item.name === '$all_event') {
            return Prisma.sql`1 = 1`;
          }

          if (item.name === '$page_view') {
            return Prisma.sql`"WebsiteEvent"."eventType" = ${EVENT_TYPE.pageView}`;
          }

          return Prisma.sql`"WebsiteEvent"."eventName" = ${item.name}`;
        }),
        ' OR ',
        '(',
        ')'
      ),
    ];
  }

  private getValueField(type: FilterInfoType): Prisma.Sql {
    const valueField =
      type === 'number'
        ? Prisma.sql`"WebsiteEventData"."numberValue"`
        : type === 'string'
          ? Prisma.sql`"WebsiteEventData"."stringValue"`
          : type === 'date'
            ? Prisma.sql`"WebsiteEventData"."dateValue"`
            : Prisma.sql`"WebsiteEventData"."numberValue"`;

    return valueField;
  }

  private buildFilterQueryOperator(
    type: FilterInfoType,
    operator: string,
    value: FilterInfoValue | null
  ) {
    const valueField = this.getValueField(type);
    return this.buildCommonFilterQueryOperator(
      type,
      operator,
      value,
      valueField
    );
  }
}

export async function insightsWebsite(
  query: z.infer<typeof insightsQuerySchema>,
  context: { timezone: string }
) {
  const { time, metrics } = query;
  const { startAt, endAt, unit, timezone = context.timezone } = time;

  const builder = new WebsiteInsightsSqlBuilder(query, context);
  const sql = builder.build();

  if (env.isDev) {
    printSQL(sql);
  }

  const res = await prisma.$queryRaw<{ date: string | null }[]>(sql);

  let result: {
    name: string;
    [groupName: string]: any;
    data: {
      date: string;
      value: number;
    }[];
  }[] = [];

  for (const m of metrics) {
    result.push({
      name: m.name,
      data: getDateArray(
        res.map((item) => {
          return {
            value: Number(get(item, m.name)),
            date: String(item.date),
          };
        }),
        startAt,
        endAt,
        unit
        // timezone // NOTICE: because return date is already with user timezone, so we don't need to pass timezone here
      ),
    });
  }

  return result;
}
