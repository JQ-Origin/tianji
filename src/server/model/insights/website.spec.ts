import { describe, expect, test } from 'vitest';
import { unwrapSQL } from '../../utils/prisma.js';
import { WebsiteInsightsSqlBuilder } from './website.js';

describe('WebsiteInsightsSqlBuilder', () => {
  const insightId = 'cly5yay7a001v5tp6xdkzmygh';
  const insightType = 'website';

  test('groups', () => {
    const builder = new WebsiteInsightsSqlBuilder(
      {
        insightId,
        insightType,
        workspaceId: '',
        metrics: [
          {
            name: '$all_event',
            math: 'events',
          },
        ],
        filters: [],
        time: {
          startAt: 1739203200000,
          endAt: 1741881599999,
          unit: 'day',
        },
        groups: [
          {
            value: 'number',
            type: 'number',
          },
        ],
      },
      {
        timezone: 'UTC',
      }
    );

    const sql = builder.build();
    expect(unwrapSQL(sql)).toMatchSnapshot('sql');
  });

  test('groups with custom bucket', () => {
    const builder = new WebsiteInsightsSqlBuilder(
      {
        insightId,
        insightType,
        workspaceId: '',
        metrics: [
          {
            name: '$all_event',
            math: 'events',
          },
        ],
        filters: [],
        time: {
          startAt: 1739203200000,
          endAt: 1741881599999,
          unit: 'day',
        },
        groups: [
          {
            value: 'number',
            type: 'number',
            customGroups: [
              {
                filterOperator: 'not equals',
                filterValue: 1,
              },
            ],
          },
        ],
      },
      {
        timezone: 'UTC',
      }
    );

    const sql = builder.build();
    expect(unwrapSQL(sql)).toMatchSnapshot('sql');
  });

  test('with filters', () => {
    const builder = new WebsiteInsightsSqlBuilder(
      {
        insightId,
        insightType,
        workspaceId: '',
        metrics: [
          {
            name: '$all_event',
            math: 'events',
          },
        ],
        filters: [
          {
            name: 'country',
            type: 'string',
            operator: 'equals',
            value: 'US',
          },
          {
            name: 'age',
            type: 'number',
            operator: 'greater than',
            value: 18,
          },
        ],
        time: {
          startAt: 1739203200000,
          endAt: 1741881599999,
          unit: 'day',
        },
        groups: [],
      },
      {
        timezone: 'UTC',
      }
    );

    const sql = builder.build();
    expect(unwrapSQL(sql)).toMatchSnapshot('sql with filters');
  });

  test('different metrics', () => {
    const builder = new WebsiteInsightsSqlBuilder(
      {
        insightId,
        insightType,
        workspaceId: '',
        metrics: [
          {
            name: '$page_view',
            math: 'events',
          },
          {
            name: 'click',
            math: 'events',
          },
          {
            name: '$all_event',
            math: 'sessions',
          },
        ],
        filters: [],
        time: {
          startAt: 1739203200000,
          endAt: 1741881599999,
          unit: 'day',
        },
        groups: [],
      },
      {
        timezone: 'UTC',
      }
    );

    const sql = builder.build();
    expect(unwrapSQL(sql)).toMatchSnapshot('sql with different metrics');
  });

  test('different time units', () => {
    const builder = new WebsiteInsightsSqlBuilder(
      {
        insightId,
        insightType,
        workspaceId: '',
        metrics: [
          {
            name: '$all_event',
            math: 'events',
          },
        ],
        filters: [],
        time: {
          startAt: 1739203200000,
          endAt: 1741881599999,
          unit: 'month',
        },
        groups: [],
      },
      {
        timezone: 'Asia/Shanghai',
      }
    );

    const sql = builder.build();
    expect(unwrapSQL(sql)).toMatchSnapshot('sql with month time unit');
  });

  test('combined filters and groups', () => {
    const builder = new WebsiteInsightsSqlBuilder(
      {
        insightId,
        insightType,
        workspaceId: '',
        metrics: [
          {
            name: '$all_event',
            math: 'events',
          },
        ],
        filters: [
          {
            name: 'isNew',
            type: 'boolean',
            operator: 'equals',
            value: 1,
          },
        ],
        time: {
          startAt: 1739203200000,
          endAt: 1741881599999,
          unit: 'day',
        },
        groups: [
          {
            value: 'browser',
            type: 'string',
          },
        ],
      },
      {
        timezone: 'UTC',
      }
    );

    const sql = builder.build();
    expect(unwrapSQL(sql)).toMatchSnapshot('sql with filters and groups');
  });

  test('date type filters', () => {
    const builder = new WebsiteInsightsSqlBuilder(
      {
        insightId,
        insightType,
        workspaceId: '',
        metrics: [
          {
            name: '$all_event',
            math: 'events',
          },
        ],
        filters: [
          {
            name: 'registeredAt',
            type: 'date',
            operator: 'between',
            value: [1739203200000, 1741881599999],
          },
        ],
        time: {
          startAt: 1739203200000,
          endAt: 1741881599999,
          unit: 'day',
        },
        groups: [],
      },
      {
        timezone: 'UTC',
      }
    );

    const sql = builder.build();
    expect(unwrapSQL(sql)).toMatchSnapshot('sql with date filters');
  });
});
