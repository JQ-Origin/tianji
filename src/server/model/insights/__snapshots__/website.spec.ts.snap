// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`WebsiteInsightsSqlBuilder > combined filters and groups > sql with filters and groups 1`] = `
"select
        to_char(date_trunc('day', "WebsiteEvent"."createdAt" at time zone 'UTC'), 'YYYY-MM-DD') date , "WebsiteEventData"."stringValue" as "%browser" , count(1) as "$all_event"
      from "WebsiteEvent" INNER JOIN "WebsiteEventData" ON "WebsiteEvent"."id" = "WebsiteEventData"."websiteEventId" AND "WebsiteEventData"."numberValue" = '1' AND "WebsiteEventData"."eventKey" = 'browser'
      where "WebsiteEvent"."websiteId" = 'cly5yay7a001v5tp6xdkzmygh' AND "WebsiteEvent"."createdAt" between '2025-02-10T16:00:00.000Z'::timestamptz and '2025-03-13T15:59:59.999Z'::timestamptz AND (1 = 1)
      group by 1 , 2
      order by 1 desc
      limit 1000"
`;

exports[`WebsiteInsightsSqlBuilder > date type filters > sql with date filters 1`] = `
"select
        to_char(date_trunc('day', "WebsiteEvent"."createdAt" at time zone 'UTC'), 'YYYY-MM-DD') date , count(1) as "$all_event"
      from "WebsiteEvent" INNER JOIN "WebsiteEventData" ON "WebsiteEvent"."id" = "WebsiteEventData"."websiteEventId" AND "WebsiteEventData"."dateValue" BETWEEN '2025-02-10T16:00:00.000Z' AND '2025-03-13T15:59:59.999Z'
      where "WebsiteEvent"."websiteId" = 'cly5yay7a001v5tp6xdkzmygh' AND "WebsiteEvent"."createdAt" between '2025-02-10T16:00:00.000Z'::timestamptz and '2025-03-13T15:59:59.999Z'::timestamptz AND (1 = 1)
      group by 1
      order by 1 desc
      limit 1000"
`;

exports[`WebsiteInsightsSqlBuilder > different metrics > sql with different metrics 1`] = `
"select
        to_char(date_trunc('day', "WebsiteEvent"."createdAt" at time zone 'UTC'), 'YYYY-MM-DD') date , sum(case WHEN "WebsiteEvent"."eventName" is null AND "WebsiteEvent"."eventType" = '1' THEN 1 ELSE 0 END) as "$page_view" , sum(case WHEN "WebsiteEvent"."eventName" = 'click' THEN 1 ELSE 0 END) as "click" , count(distinct "sessionId") as "$all_event"
      from "WebsiteEvent" 
      where "WebsiteEvent"."websiteId" = 'cly5yay7a001v5tp6xdkzmygh' AND "WebsiteEvent"."createdAt" between '2025-02-10T16:00:00.000Z'::timestamptz and '2025-03-13T15:59:59.999Z'::timestamptz AND ("WebsiteEvent"."eventType" = '1' OR "WebsiteEvent"."eventName" = 'click' OR 1 = 1)
      group by 1
      order by 1 desc
      limit 1000"
`;

exports[`WebsiteInsightsSqlBuilder > different time units > sql with month time unit 1`] = `
"select
        to_char(date_trunc('month', "WebsiteEvent"."createdAt" at time zone 'UTC'), 'YYYY-MM-01') date , count(1) as "$all_event"
      from "WebsiteEvent" 
      where "WebsiteEvent"."websiteId" = 'cly5yay7a001v5tp6xdkzmygh' AND "WebsiteEvent"."createdAt" between '2025-02-10T16:00:00.000Z'::timestamptz and '2025-03-13T15:59:59.999Z'::timestamptz AND (1 = 1)
      group by 1
      order by 1 desc
      limit 1000"
`;

exports[`WebsiteInsightsSqlBuilder > groups > sql 1`] = `
"select
        to_char(date_trunc('day', "WebsiteEvent"."createdAt" at time zone 'UTC'), 'YYYY-MM-DD') date , "WebsiteEventData"."numberValue" as "%number" , count(1) as "$all_event"
      from "WebsiteEvent" INNER JOIN "WebsiteEventData" ON "WebsiteEvent"."id" = "WebsiteEventData"."websiteEventId" AND "WebsiteEventData"."eventKey" = 'number'
      where "WebsiteEvent"."websiteId" = 'cly5yay7a001v5tp6xdkzmygh' AND "WebsiteEvent"."createdAt" between '2025-02-10T16:00:00.000Z'::timestamptz and '2025-03-13T15:59:59.999Z'::timestamptz AND (1 = 1)
      group by 1 , 2
      order by 1 desc
      limit 1000"
`;

exports[`WebsiteInsightsSqlBuilder > groups with custom bucket > sql 1`] = `
"select
        to_char(date_trunc('day', "WebsiteEvent"."createdAt" at time zone 'UTC'), 'YYYY-MM-DD') date , "WebsiteEventData"."numberValue" != '1' as "%number|not equals|1" , count(1) as "$all_event"
      from "WebsiteEvent" INNER JOIN "WebsiteEventData" ON "WebsiteEvent"."id" = "WebsiteEventData"."websiteEventId" AND "WebsiteEventData"."eventKey" = 'number'
      where "WebsiteEvent"."websiteId" = 'cly5yay7a001v5tp6xdkzmygh' AND "WebsiteEvent"."createdAt" between '2025-02-10T16:00:00.000Z'::timestamptz and '2025-03-13T15:59:59.999Z'::timestamptz AND (1 = 1)
      group by 1 , 2
      order by 1 desc
      limit 1000"
`;

exports[`WebsiteInsightsSqlBuilder > with filters > sql with filters 1`] = `
"select
        to_char(date_trunc('day', "WebsiteEvent"."createdAt" at time zone 'UTC'), 'YYYY-MM-DD') date , count(1) as "$all_event"
      from "WebsiteEvent" INNER JOIN "WebsiteEventData" ON "WebsiteEvent"."id" = "WebsiteEventData"."websiteEventId" AND "WebsiteEventData"."stringValue" = 'US' AND "WebsiteEventData"."numberValue" > '18'
      where "WebsiteEvent"."websiteId" = 'cly5yay7a001v5tp6xdkzmygh' AND "WebsiteEvent"."createdAt" between '2025-02-10T16:00:00.000Z'::timestamptz and '2025-03-13T15:59:59.999Z'::timestamptz AND (1 = 1)
      group by 1
      order by 1 desc
      limit 1000"
`;
